test/integration/smoke/test_host_maintenance.py

```diff
@@ -290,3 +274,313 @@ def test_02_cancel_host_maintenace_with_migration_jobs(self):
+            )
+            if isinstance(response, list):
+                if response[0].state == state:
+                    self.logger.debug('Host agent with id %s is in state = %s' % (hostid, state))
+                    return True, None
+            return False, None
+
+        done, _ = wait_until(interval, retries, check_agent_state)
+        if not done:
+            raise Exception("Failed to wait for host agent %s to be on state %s" % (hostid, state))
+        return True
+
+    def cancel_host_maintenance(self, hostid):
+        cmd = cancelHostMaintenance.cancelHostMaintenanceCmd()
+        cmd.id = hostid
+        self.apiclient.cancelHostMaintenance(cmd)
+        self.logger.debug('Host with id %s is cancelling maintenance' % hostid)
+
+    def get_enabled_host_connected_agent(self):
+        hosts = Host.list(
+            self.apiclient,
+            type='Routing',
+            zoneid=self.zone.id,
+            podid=self.pod.id,
+            hypervisor=self.hypervisor,
+            resourcestate='Enabled',
+            state='Up'
+        )
+        if len(hosts) < 2:
+            raise unittest.SkipTest("Cancel host maintenance must be tested for 2 or more hosts")
+        return hosts[0]
+
+    def deploy_vm_on_host(self, hostid):
+        return VirtualMachine.create(
+            self.apiclient,
+            self.services["virtual_machine"],
+            serviceofferingid=self.service_offering.id,
+            networkids=self.l2_network.id,
+            hostid=hostid
+        )
+
+    def assert_host_is_functional_after_cancelling_maintenance(self, hostid):
+        self.wait_until_agent_is_in_state(hostid, "Up")
+        self.logger.debug('Deploying VM on host %s' % hostid)
+        vm = self.deploy_vm_on_host(hostid)
+        self.assertEqual(
+            vm.state,
+            "Running",
+            "Check VM is running on the host"
+        )
+        self.cleanup.append(vm)
+
+    @skipTestIf("hypervisorNotSupported")
+    @attr(tags=["advanced", "advancedns", "smoke", "basic", "eip", "sg"], required_hardware="true")
+    def test_01_cancel_host_maintenance_ssh_enabled_agent_connected(self):
+        """
+        Test cancel maintenance when: 'kvm.ssh.to.agent' = true, agent state = 'Up'
+
+        1) Put host on Maintenance
+        2) Cancel maintenance on host
+        4) Assert agent is still connected after cancelling maintenance
+        3) Deploy VM on the host after cancelling maintenance
+        """
+
+        if not self.is_ssh_enabled():
+            self.set_ssh_enabled(True)
+
+        try:
+            self.prepare_host_for_maintenance(self.host.id)
+            self.wait_until_host_is_in_state(self.host.id, "Maintenance")
+            self.cancel_host_maintenance(self.host.id)
+            self.wait_until_host_is_in_state(self.host.id, "Enabled")
+            self.assert_host_is_functional_after_cancelling_maintenance(self.host.id)
+        except Exception as e:
+            self.fail(e)
```

Comment:
I think it's vital to have a call to a method here that will iterate through all hosts and set them all back to Up. If there's an error within this 'try' it will leave the host into maintenance and the rest of the test run will be doomed.